storage: fix Store.haveGossiped{SystemConfig,NodeLiveness} logic

petermattis opened the pull request:
I still need to add a test that shows the failure to gossip the system
config with the previous logic.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/13167)
<!-- Reviewable:end -->

petermattis committed:
config,keys: force a split at the start of the system-config span

Force a split at the start of the system-config span. This places the
system-config in its own range and prevents the oddity of the
system-config span starting off in a range with an expiration-based
range lease and then moving to a range with an epoch-based range lease.

Placing the system-config in its own span also revealed a lurking bug in
processing of the system-config trigger. All of the SQL DDL
operations (e.g. CREATE DATABASE) increment /system/desc-idgen as the
first write of the transaction which anchors the transaction to the
range containing that key. But we need the transaction anchored to the
range containing the system-config span. We now make this explicit when
generating the BeginTransactionRequest if the system-config trigger has
been set.

petermattis committed:
storage: fix Store.haveGossiped{SystemConfig,NodeLiveness} logic

Add a retry loop to periodic gossips.

tamird commented:
LGTM (with test) but also seems like we could just get rid of these atomics and compare the new lease to the old lease more deeply (e.g. by also considering epoch).

---

Reviewed 9 of 9 files at r1, 2 of 2 files at r2.
Review status: 6 of 11 files reviewed at latest revision, 5 unresolved discussions.

---

*[pkg/config/config.go, line 374 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13167#-KbQqo3Y2IsWqgatuOXZ:-KbQqo3Y2IsWqgatuOX_:btfibz1) ([raw file](https://github.com/cockroachdb/cockroach/blob/4aa9bb2ab36db6b61a4dad3715930dbaca54d3e9/pkg/config/config.go#L374)):*
> ```Go
> // ComputeSplitKeys takes a start and end key and returns an array of keys
> // at which to split the span [start, end).
> // The only required splits are at each user table prefix.
> ```

this comment is outdated. perhaps it's time to replace this comment and the comments in the beginning of this function with a simple diagram

---

*[pkg/internal/client/txn.go, line 710 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13167#-KbQr7bSGHcsMfaVlBfD:-KbQr7bSGHcsMfaVlBfE:bk89uph) ([raw file](https://github.com/cockroachdb/cockroach/blob/4aa9bb2ab36db6b61a4dad3715930dbaca54d3e9/pkg/internal/client/txn.go#L710)):*
> ```Go
> 	// begin transaction request before the first write command.
> 	if needBeginTxn {
> 		// If we're setting the system-config trigger, anchor the transaction to
> ```

this comment should explain why this is done; currently, it is just repeating the code.

---

*[pkg/internal/client/txn.go, line 713 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13167#-KbQroNgXbULt-u8yomK:-KbQroNgXbULt-u8yomL:b-hss2ba) ([raw file](https://github.com/cockroachdb/cockroach/blob/4aa9bb2ab36db6b61a4dad3715930dbaca54d3e9/pkg/internal/client/txn.go#L713)):*
> ```Go
> 		// the system-config range.
> 		if txn.systemConfigTrigger {
> 			firstWriteKey = keys.SystemConfigSpan.Key
> ```

might be good to rename this variable, since it's no longer the first write key in all cases.

---

*[pkg/keys/constants.go, line 237 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13167#-KbQrGn7blt9nFOfdkrp:-KbQrGn7blt9nFOfdkrq:btoxfsd) ([raw file](https://github.com/cockroachdb/cockroach/blob/4aa9bb2ab36db6b61a4dad3715930dbaca54d3e9/pkg/keys/constants.go#L237)):*
> ```Go
> 
> 	// TableDataMin is the start of the range of table data keys.
> 	TableDataMin = roachpb.Key(encoding.EncodeVarintAscending(nil, 0))
> ```

why'd this change?

---

*[pkg/keys/spans.go, line 30 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13167#-KbQrOAxilMupi_EB_7H:-KbQrOAxilMupi_EB_7I:b-l5mo2p) ([raw file](https://github.com/cockroachdb/cockroach/blob/4aa9bb2ab36db6b61a4dad3715930dbaca54d3e9/pkg/keys/spans.go#L30)):*
> ```Go
> 	// SystemConfigSpan is the range of system objects which will be gossiped.
> 	SystemConfigSpan = roachpb.Span{
> 		Key:    MakeRowSentinelKey(TableDataMin),
> ```

this is a departure from other keys here; comment?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13167#-:-KbQsKLO5M40HamJQlDx:b-mb7en7)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 3 of 5 files at r3, 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[pkg/config/config.go, line 387 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13167#-KbQsVkJ2JOfHMrchADf:-KbQsVkKvvOrAvlzu6Ri:bly58ea) ([raw file](https://github.com/cockroachdb/cockroach/blob/a366f4313e09924e793b0353f91e4902cb5aa00e/pkg/config/config.go#L387)):*
> ```Go
> 			return nil
> 		}
> 		// NB: Split keys need to be valid column keys.
> ```

this comment should be on `keys.SystemConfigSplitKey`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13167)*
<!-- Sent from Reviewable.io -->

spencerkimball commented on pkg/internal/client/txn.go:
```diff
@@ -706,6 +707,11 @@ func (txn *Txn) send(ba roachpb.BatchRequest) (*roachpb.BatchResponse, *roachpb.
 	// If we're not yet writing in this txn, but intend to, insert a
 	// begin transaction request before the first write command.
 	if needBeginTxn {
+		// If we're setting the system-config trigger, anchor the transaction to
```
This doesn't seem like the right place to put this logic. The logic around system-config is too specific for `internal/client`. Instead, the code which creates the transaction to modify system-config should make sure the first key modified is the system config span. This is similar to what we do when we split or merge ranges.

petermattis replied:
That is very fragile. Please take a look at what you're suggesting (i.e. try and figure out what we would need to modify in the `sql` package). I did consider adding a `Txn.SetTxnKey()` method so that the user could specify how they wanted a transaction anchor, but that felt overly specific too. If we require the system-config trigger to be associated with a transaction anchored on the system-config range, why not make that automatic?

spencerkimball replied:
I don't expect that knowledge of system config details should leak into this high level client code. I'd be much happier with `Txn.SetTxnKey`.

petermattis replied:
We already have `txn.SetSystemConfigTrigger`. Why do we need a separate call to set the txn key for such transactions?

spencerkimball replied:
Because they're separate concepts. I'd like to use `Txn.SetTxnKey` to make splitting, merging and change replicas less fragile.

a-robinson commented:
Reviewed 6 of 9 files at r1, 5 of 5 files at r3, 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica.go, line 4376 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13167#-KbQxHq4xMO77ZFdNR2K:-KbQxHq4xMO77ZFdNR2L:b-uovq5p) ([raw file](https://github.com/cockroachdb/cockroach/blob/a366f4313e09924e793b0353f91e4902cb5aa00e/pkg/storage/replica.go#L4376)):*
> ```Go
> 
> 	// log.VEventf(ctx, 2, "gossiping system config")
> 	log.Infof(ctx, "gossiping system config")
> ```

Need to clean up your debug logging

---

*[pkg/storage/replica.go, line 4431 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13167#-KbQxVeEjjOMkNtRj6Dj:-KbQxVeEjjOMkNtRj6Dk:b-mur16l) ([raw file](https://github.com/cockroachdb/cockroach/blob/a366f4313e09924e793b0353f91e4902cb5aa00e/pkg/storage/replica.go#L4431)):*
> ```Go
> 		}
> 	}
> 	atomic.StoreInt32(&r.store.haveGossipedNodeLiveness, 1)
> ```

This isn't totally true if any of the iterations of the loop above ran into an error. Right now it looks like the errors will only happen in cases where retrying won't help, but it's certainly possible that more error cases will be added inside the loop or the methods it calls.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13167)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Added a test which fails reliably before this change. I also had to add in the retry-loop for the periodic gossip in order to remove flakiness from the test. See the comment in the retry loop as to why this was necessary (we may not be able to get an epoch-based lease the first time we try to gossip the system config).

PTAL

---

Review status: 6 of 14 files reviewed at latest revision, 9 unresolved discussions.

---

*[pkg/config/config.go, line 374 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13167#-KbQqo3Y2IsWqgatuOXZ:-KbR6pl20IFAuuSLlwXt:by960dk) ([raw file](https://github.com/cockroachdb/cockroach/blob/4aa9bb2ab36db6b61a4dad3715930dbaca54d3e9/pkg/config/config.go#L374)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

this comment is outdated. perhaps it's time to replace this comment and the comments in the beginning of this function with a simple diagram
</blockquote></details>

Updated the comment.

---

*[pkg/config/config.go, line 387 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13167#-KbQsVkJ2JOfHMrchADf:-KbR6uNK64swsCoLhFl6:bwhglds) ([raw file](https://github.com/cockroachdb/cockroach/blob/a366f4313e09924e793b0353f91e4902cb5aa00e/pkg/config/config.go#L387)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

this comment should be on `keys.SystemConfigSplitKey`?
</blockquote></details>

Moved.

---

*[pkg/internal/client/txn.go, line 710 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13167#-KbQr7bSGHcsMfaVlBfD:-KbR7S6xoYIO9s-tuLN3:b-jejdf1) ([raw file](https://github.com/cockroachdb/cockroach/blob/4aa9bb2ab36db6b61a4dad3715930dbaca54d3e9/pkg/internal/client/txn.go#L710)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

this comment should explain why this is done; currently, it is just repeating the code.
</blockquote></details>

Added a comment to `SetSystemConfigTrigger`.

---

*[pkg/internal/client/txn.go, line 713 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13167#-KbQroNgXbULt-u8yomK:-KbR7doqOl4FZhLBHH3G:b-gqt88o) ([raw file](https://github.com/cockroachdb/cockroach/blob/4aa9bb2ab36db6b61a4dad3715930dbaca54d3e9/pkg/internal/client/txn.go#L713)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

might be good to rename this variable, since it's no longer the first write key in all cases.
</blockquote></details>

Renamed to `txnAnchorKey`.

---

*[pkg/internal/client/txn.go, line 710 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13167#-KbQoGN54BLBNWmN1l8T-r4-710:-KbR7LXeYhu00LzNfDBI:b-le6hh3) ([raw file](https://github.com/cockroachdb/cockroach/blob/a366f4313e09924e793b0353f91e4902cb5aa00e/pkg/internal/client/txn.go#L710)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary><blockquote>

Because they're separate concepts. I'd like to use `Txn.SetTxnKey` to make splitting, merging and change replicas less fragile.
</blockquote></details>

I've compromised and added `SetTxnAnchorKey` but made `SetSystemConfigTrigger` call it.

---

*[pkg/keys/constants.go, line 237 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13167#-KbQrGn7blt9nFOfdkrp:-KbR7fvoltbvVGOz2wPo:b-ofho4k) ([raw file](https://github.com/cockroachdb/cockroach/blob/4aa9bb2ab36db6b61a4dad3715930dbaca54d3e9/pkg/keys/constants.go#L237)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

why'd this change?
</blockquote></details>

`math.MinInt64` is a very larger negative number which was confusing in logs. Note that we weren't previously doing anything interesting with this constant.

---

*[pkg/storage/replica.go, line 4376 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13167#-KbQxHq4xMO77ZFdNR2K:-KbR8As-jxLMdzrzcGpS:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a366f4313e09924e793b0353f91e4902cb5aa00e/pkg/storage/replica.go#L4376)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

Need to clean up your debug logging
</blockquote></details>

Done.

---

*[pkg/storage/replica.go, line 4431 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13167#-KbQxVeEjjOMkNtRj6Dj:-KbR8OlyQI0SpYdr3Kdn:b-khym0o) ([raw file](https://github.com/cockroachdb/cockroach/blob/a366f4313e09924e793b0353f91e4902cb5aa00e/pkg/storage/replica.go#L4431)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

This isn't totally true if any of the iterations of the loop above ran into an error. Right now it looks like the errors will only happen in cases where retrying won't help, but it's certainly possible that more error cases will be added inside the loop or the methods it calls.
</blockquote></details>

Hrmm, I don't think this is causing a problem right now. Node liveness is a lot different from the system-config in that we're gossiping it very frequently (every 2-3 secs). We could probably get rid of the gossiping of node liveness when the lease changes hands. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13167#-:-KbR8kQ-zMskX_wJv1M7:b-rp85nn)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 4 of 6 files at r5, 4 of 4 files at r6.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

*[pkg/config/config.go, line 378 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/13167#-KbR9CtOE6imM6iGsJJQ:-KbR9CtOE6imM6iGsJJR:bpkvyht) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac0b812764d15b162b69f217e87330a12d4eac37/pkg/config/config.go#L378)):*
> ```Go
> // of the system-config range (i.e. /table/0). The system-config range is
> // somewhat special in that it contains multiple SQL tables
> // (/table/0-/table/10) all of which must be contained in a single range.
> ```

s/all of//

might also be good to avoid the literal "10" here and instead reference a constant

---

*[pkg/storage/client_lease_test.go, line 189 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/13167#-KbRBDjMHvueYCwQdYup:-KbRBDjMHvueYCwQdYuq:bjnzgwf) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac0b812764d15b162b69f217e87330a12d4eac37/pkg/storage/client_lease_test.go#L189)):*
> ```Go
> 
> 	getSystemConfig := func() config.SystemConfig {
> 		systemConfig, ok := mtc.gossips[0].GetSystemConfig()
> ```

isn't this equivalent to `sc, _ : = ...GetSystemConfig(); return sc`?

---

*[pkg/storage/replica.go, line 4431 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13167#-KbQxVeEjjOMkNtRj6Dj:-KbRBazW1-pstkohhKoa:b-rwhkt5) ([raw file](https://github.com/cockroachdb/cockroach/blob/a366f4313e09924e793b0353f91e4902cb5aa00e/pkg/storage/replica.go#L4431)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Hrmm, I don't think this is causing a problem right now. Node liveness is a lot different from the system-config in that we're gossiping it very frequently (every 2-3 secs). We could probably get rid of the gossiping of node liveness when the lease changes hands. 
</blockquote></details>

+1, we should file an issue to do that.

---

*[pkg/storage/store.go, line 1326 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/13167#-KbRATWPs3OBTChg9G5u:-KbRATWQrAQnjQFKd_g2:brmjl60) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac0b812764d15b162b69f217e87330a12d4eac37/pkg/storage/store.go#L1326)):*
> ```Go
> 				retryOptions := base.DefaultRetryOptions()
> 				retryOptions.Closer = s.stopper.ShouldStop()
> 				for r := retry.Start(retryOptions); r.Next(); {
> ```

this code still has some idiosyncrasies that I tried to address in #13147; which part of that change to you object to?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13167)*
<!-- Sent from Reviewable.io -->

a-robinson commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 3 of 6 files at r5, 4 of 4 files at r6.
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13167#-:-KbRKBTxxRbQqmP_gOxG:bnfp4nl)*
<!-- Sent from Reviewable.io -->

spencerkimball commented on pkg/internal/client/txn.go:
```diff
@@ -147,13 +151,23 @@ func (txn *Txn) InternalSetPriority(priority int32) {
+// SetTxnAnchorKey sets the key at which to anchor the transaction record. The
+// transaction anchor key defaults to the first key written in a transaction.
+func (txn *Txn) SetTxnAnchorKey(key roachpb.Key) {
+	txn.txnAnchorKey = key
```
We must panic here if the transaction has already had `BeginTransaction` invoked.

petermattis commented:
Review status: 12 of 15 files reviewed at latest revision, 7 unresolved discussions.

---

*[pkg/config/config.go, line 378 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/13167#-KbR9CtOE6imM6iGsJJQ:-KbRXzrMdDbOm2IO6XNx:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac0b812764d15b162b69f217e87330a12d4eac37/pkg/config/config.go#L378)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

s/all of//

might also be good to avoid the literal "10" here and instead reference a constant
</blockquote></details>

Done.

---

*[pkg/internal/client/txn.go, line 168 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/13167#-KbQoGN54BLBNWmN1l8T-r6-168:-KbRYTH-OVe-5sMK4XO1:b-d7nko8) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac0b812764d15b162b69f217e87330a12d4eac37/pkg/internal/client/txn.go#L168)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary><blockquote>

We must panic here if the transaction has already had `BeginTransaction` invoked.
</blockquote></details>

Done. This found a bug in `sql.SchemaChanger`.

---

*[pkg/storage/client_lease_test.go, line 189 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/13167#-KbRBDjMHvueYCwQdYup:-KbRZM89nekWkLQRt_zP:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac0b812764d15b162b69f217e87330a12d4eac37/pkg/storage/client_lease_test.go#L189)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

isn't this equivalent to `sc, _ : = ...GetSystemConfig(); return sc`?
</blockquote></details>

Done.

---

*[pkg/storage/replica.go, line 4431 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13167#-KbQxVeEjjOMkNtRj6Dj:-KbRZMt9WSHPhy2_DUfQ:bmfck81) ([raw file](https://github.com/cockroachdb/cockroach/blob/a366f4313e09924e793b0353f91e4902cb5aa00e/pkg/storage/replica.go#L4431)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

+1, we should file an issue to do that.
</blockquote></details>

I'll file an issue.

---

*[pkg/storage/store.go, line 1326 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/13167#-KbRATWPs3OBTChg9G5u:-KbRZOKJEo5CznZvJj_g:bm512na) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac0b812764d15b162b69f217e87330a12d4eac37/pkg/storage/store.go#L1326)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

this code still has some idiosyncrasies that I tried to address in #13147; which part of that change to you object to?
</blockquote></details>

I didn't object to anything specific in #13147, just that it was making changes in the absence of understanding the problem. Let's revisit it once this is merged.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13167)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 3 of 3 files at r7.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---

*[pkg/storage/client_lease_test.go, line 189 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/13167#-KbRBDjMHvueYCwQdYup:-KbRecBfyw5xhny9PES5:b5maicy) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac0b812764d15b162b69f217e87330a12d4eac37/pkg/storage/client_lease_test.go#L189)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

What's done?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13167)*
<!-- Sent from Reviewable.io -->

petermattis committed:
storage: potentially gossip system-data on every lease acquisition

Remove Store.haveGossiped{SystemConfig,NodeLiveness}. We now gossip the
system-config and node-liveness on every lease acquisition. This is not
overly extensive due to the frequent gossiping of node liveness in any
case and the infrequent reacquisition of the system config range lease.

petermattis commented:
Added another commit which gets rid of `Store.haveGossiped...` in favor of gossiping whenever the lease epoch changes. Simpler approach and the new test still passes.

---

Review status: 11 of 15 files reviewed at latest revision, 5 unresolved discussions.

---

*[pkg/storage/client_lease_test.go, line 189 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/13167#-KbRBDjMHvueYCwQdYup:-KbRfs2jm-aa9ayl7nQw:b-iep4b8) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac0b812764d15b162b69f217e87330a12d4eac37/pkg/storage/client_lease_test.go#L189)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

What's done?
</blockquote></details>

Lost in my `git stash`. Look again. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13167#-:-KbRfx5Y9_5u8jHqydwV:b-7n0jah)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r8, 3 of 3 files at r9.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

*[pkg/storage/client_lease_test.go, line 188 at r9](https://reviewable.io:443/reviews/cockroachdb/cockroach/13167#-KbRgPNSUu76sK0IrQbl:-KbRgPNSUu76sK0IrQbm:bv2t74) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ebcf75a5f8a7e8ef5aea1448bd0f096ea69a94c/pkg/storage/client_lease_test.go#L188)):*
> ```Go
> 	mtc.stopStore(0)
> 
> 	getSystemConfig := func() config.SystemConfig {
> ```

btw, it seems that all you care about from these functions is whether or not the return value is empty, so both could be reduced to returning bools

---

*[pkg/storage/replica_proposal.go, line 444 at r9](https://reviewable.io:443/reviews/cockroachdb/cockroach/13167#-KbRgXXmnf454sviShgu:-KbRgXXmnf454sviShgv:b-18lal) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ebcf75a5f8a7e8ef5aea1448bd0f096ea69a94c/pkg/storage/replica_proposal.go#L444)):*
> ```Go
> 		// reacquires a lease it previously held). Note that Lease.Epoch is 0 for
> 		// expiration-based leases.
> 		if leaseChangingHands || prevLease.Epoch != newLease.Epoch {
> ```

commented on the commit, but reviewable didn't notice. this is comparing pointers.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13167)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 14 of 15 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[pkg/storage/client_lease_test.go, line 188 at r9](https://reviewable.io:443/reviews/cockroachdb/cockroach/13167#-KbRgPNSUu76sK0IrQbl:-KbRp1hKsATGSkeVhqRY:bhwaq80) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ebcf75a5f8a7e8ef5aea1448bd0f096ea69a94c/pkg/storage/client_lease_test.go#L188)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

btw, it seems that all you care about from these functions is whether or not the return value is empty, so both could be reduced to returning bools
</blockquote></details>

True, but that seems less obvious to me. I'd have to rename these functions to `get{SystemConfig,NodeLiveness}NonEmpty`. I'm going to leave as-is for now.

---

*[pkg/storage/replica_proposal.go, line 444 at r9](https://reviewable.io:443/reviews/cockroachdb/cockroach/13167#-KbRgXXmnf454sviShgu:-KbRq8Qhcsi1uKZ0CNlM:b-zh2yfp) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ebcf75a5f8a7e8ef5aea1448bd0f096ea69a94c/pkg/storage/replica_proposal.go#L444)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

commented on the commit, but reviewable didn't notice. this is comparing pointers.
</blockquote></details>

Yep. Fixed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13167)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r10.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13167)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 13 of 15 files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---

*[pkg/storage/replica_proposal.go, line 451 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/13167#-KbSoot9Bif_Ti2FfIZr:-KbSoot9Bif_Ti2FfIZs:b-klku92) ([raw file](https://github.com/cockroachdb/cockroach/blob/24523fe75030f4039d2eb582fee16aaf21cc7e90/pkg/storage/replica_proposal.go#L451)):*
> ```Go
> 			newEpoch = *newLease.Epoch
> 		}
> 		if leaseChangingHands || prevEpoch != newEpoch {
> ```

Turns out this wasn't kosher during bootstrapping. A lease on `r1` would be acquired but we wouldn't be able to gossip the system-config because gossip is not set up during bootstrapping. If we then went right to initialization, the same lease is used and we never gossip the system-config. See the new comment about why all this complexity was misplaced. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13167)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 2 files at r11.
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13167)*
<!-- Sent from Reviewable.io -->
